src/nearest_neighbors/datasets/synthetic_data/loader.py

```diff
@@ -186,8 +187,90 @@ def _make_mcar(self) -> None:
         self.availability_mask = A
 
     def _make_mnar(self) -> None:
-        raise NotImplementedError("MNAR yet to be implemented")
-        # TODO: Aashish/Caleb/Kyuesong/Tatha: come back to this later
+        """Makes values missing not at random (MNAR), specifically staggered adoption (non-positivity & confounded)"""
+        U = self.row_latent
+
+        N = self.num_rows
+        T = self.num_cols
+
+        missing_mask = np.zeros((N, T))
+        pre_Masking = np.zeros((N, T))
+
+        # Divide units into 3 groups
+        g1_inds = np.arange(0, N // 3)
+        g2_inds = np.arange(N // 3, 2 * N // 3)
+        g3_inds = np.arange(2 * N // 3, N)
+
+        gamma_1 = [2, 0.7, 1, 0.7]
+        gamma_2 = [2, 0.2, 1, 0.2]
+
+        # TODO: make beta a parameter (currently hardcoded)
+        # first group adopts at the first 30% of the time period
+        # second group adopts at the first 70% of the time period
+        beta = [0.3, 0.7]
+
+        T1_lower = math.floor(T ** beta[0])
+        T2_lower = math.floor(T ** beta[1])
+
+        for i in range(N):
+            if i in g1_inds:
+                pre_Masking[i, :] = np.concatenate(
+                    (np.ones(T1_lower), np.zeros(T - T1_lower))
+                )
+                for t in range(T - T1_lower):
+                    pre_Masking[i, (t + T1_lower)] = (
+                        np.random.binomial(  # each units' adoption time probability is affected by their neighbors
+                            1,
+                            self._expit(
+                                gamma_1[0]
+                                + (0.99**t) * gamma_1[1] * U[i - 1]
+                                + gamma_1[2] * U[i]
+                                + (0.99**t) * gamma_1[3] * U[i + 1]
+                            ),
+                            1,
+                        )
+                    )
+                pre_A = pre_Masking[i, :]
+                if len([i for i in range(len(pre_A)) if pre_A[i] == 0]) == 0:
+                    missing_mask[i, :] = pre_A
+                elif len([i for i in range(len(pre_A)) if pre_A[i] == 0]) > 0:
+                    adopt_time = min([i for i in range(len(pre_A)) if pre_A[i] == 0])
+                    missing_mask[i, :] = np.concatenate(
+                        (np.ones(adopt_time), np.zeros(T - adopt_time))
+                    )
+            elif i in g2_inds:
+                pre_Masking[i, :] = np.concatenate(
+                    (np.ones(T2_lower), np.zeros(T - T2_lower))
+                )
+                for t in range(T - T2_lower):
+                    pre_Masking[i, (t + T2_lower)] = np.random.binomial(
+                        1,
+                        self._expit(
+                            gamma_2[0]
+                            + (1.01**t) * gamma_2[1] * U[i - 1]
```

Comment:
same issue as 0.99. Maybe it should be 2 - new_const_name? What is this for anyway?